File: frontend/src/components/sqlconstrutor/insere_servico.js
export default function insere_servico(inputText) {
  let query = "INSERT INTO Servicos\n (prazo_guarda_interna,\n admite_residuo,\n tentativas_externas_previstas,\n sigla,\n descricao,\n categoria,\n familia,\n gera_pre_alerta,\n hora_real_entrega,\n dados_do_recebedor_na_baixa,\n entrega_externa,\n entrega_com_imagem) VALUES\n";

  for (let linha of inputText.split("\n")) {
    if (linha.trim() === "") continue;

    let campos = linha.split("\t");
    let valores = campos.slice(0, 10).map(valor => isNaN(valor) ? "'" + valor + "'" : valor).join(", ");
    valores += ", " + campos.slice(10).join(", ");
    query += "(" + valores + "),\n";
  }

  return query.slice(0, -2) + "ON DUPLICATE KEY UPDATE prazo_guarda_interna = VALUES(prazo_guarda_interna), admite_residuo = VALUES(admite_residuo), tentativas_externas_previstas = VALUES(tentativas_externas_previstas), sigla = VALUES(sigla), descricao = VALUES(descricao), categoria = VALUES(categoria), familia = VALUES(familia), gera_pre_alerta = VALUES(gera_pre_alerta), hora_real_entrega = VALUES(hora_real_entrega), dados_do_recebedor_na_baixa = VALUES(dados_do_recebedor_na_baixa), entrega_externa = VALUES(entrega_externa), entrega_com_imagem = VALUES(entrega_com_imagem);";
}

File: sge_api/src/controllers/insere_servico.py
from flask import request
from flask_restx import Resource
from flask_restx import fields
from sqlalchemy import text, exists
from src.server.instance import server
from src.database.sge import db
from src.models.flask_restx_models import servicos_flask_restx_model
from src.models.Servicos_sqlalchemy_model import Servicos_sqlalchemy_model
import json
from datetime import datetime

app, api = server.app, server.api

def datetime_handler(obj):
    if isinstance(obj, datetime):
        return obj.isoformat()
    raise TypeError(f'Object of type {obj.__class__.__name__} is not JSON serializable')

@api.route('/insere_servico')
class InsereServico(Resource):
    @api.expect(servicos_flask_restx_model)
    def post(self):
        input_data = request.get_json()['data']
        rows = input_data.split('\n')
        print(input_data)
        data = []
        registros_inseridos = []
        for row in rows:
            if row:
                values = row.split('\t')
                sigla = values[3]
                if db.session.query(exists().where(Servicos_sqlalchemy_model.sigla == sigla)).scalar():
                    continue
                registros_inseridos.append(sigla)
                data.append({
                    'prazo_guarda_interna': int(values[0]),
                    'admite_residuo': bool(int(values[1])),
                    'tentativas_externas_previstas': int(values[2]),
                    'sigla': sigla,
                    'descricao': values[4],
                    'categoria': values[5],
                    'familia': values[6],
                    'gera_pre_alerta': bool(int(values[7])),
                    'hora_real_entrega': bool(int(values[8])),
                    'dados_do_recebedor_na_baixa': values[9],
                    'entrega_externa': bool(int(values[10])),
                    'entrega_com_imagem': bool(int(values[11])),
                })
        db.session.bulk_insert_mappings(Servicos_sqlalchemy_model, data)
        db.session.commit()

        # Consulta os registros onde a sigla está presente na lista registros_inseridos
        registros = Servicos_sqlalchemy_model.query.filter(Servicos_sqlalchemy_model.sigla.in_(registros_inseridos)).all()
        # Retorna os registros para o frontend em formato JSON
        return json.dumps({'registros': [r.to_dict() for r in registros]}, default=datetime_handler)